lib/notifications/notification.ex

```diff
@@ -92,61 +98,101 @@ defmodule Notifications.Notification do
-                route_ids: bw.route_ids,
-                run_ids: bw.run_ids,
-                trip_ids: bw.trip_ids,
-                operator_id: bw.operator_id,
-                operator_name: bw.operator_name,
-                route_id_at_creation: bw.route_id_at_creation,
-                block_id: bw.block_id,
-                service_id: bw.service_id,
-                start_time: bw.start_time,
-                end_time: bw.end_time
-              },
               where:
-                bw.start_time == ^notification_without_id.start_time and
-                  bw.end_time == ^notification_without_id.end_time and
-                  bw.block_id == ^notification_without_id.block_id and
-                  bw.service_id == ^notification_without_id.service_id and
-                  bw.reason == ^notification_without_id.reason
+                bw.start_time == ^block_waiver_values.start_time and
+                  bw.end_time == ^block_waiver_values.end_time and
+                  bw.block_id == ^block_waiver_values.block_id and
+                  bw.service_id == ^block_waiver_values.service_id and
+                  bw.reason == ^block_waiver_values.reason
+            )
+          )
+      end
+
+    struct!(__MODULE__, Map.merge(notification_values, %{id: db_record.id}))
+  end
+
+  def get_or_create_from_bridge_movement(bridge_movement_values) do
+    created_at = DateTime.utc_now() |> DateTime.to_unix()
+
+    {:ok, {source, db_record}} =
+      Skate.Repo.transaction(fn ->
+        # The bridge API doesn't have anything like an "updated at" timestamp
+        # so we assume that, if we see two bridge movements within a blackout
+        # period, they are actually the same movement and what's happening is
+        # that we have multiple servers trying to insert a movement at once.
+        cutoff_time = NaiveDateTime.utc_now() |> NaiveDateTime.add(-@bridge_lowering_blackout)
+
+        existing_record =
+          Skate.Repo.one(
+            from(n in DbNotification,
+              join: bm in assoc(n, :bridge_movement),
+              where: bm.inserted_at > ^cutoff_time,
+              order_by: [desc: bm.inserted_at],
+              limit: 1
             )
           )
+
+        if existing_record do
+          {:existing_record, existing_record}
+        else
+          changeset =
+            DbNotification.bridge_movement_changeset(
+              %DbNotification{created_at: created_at},
+              bridge_movement_values
+            )
+
+          {:ok, new_record} = Skate.Repo.insert(changeset)
+          link_notification_to_users(new_record.id, @chelsea_st_bridge_route_ids)
+          {:new_record, new_record}
+        end
+      end)
+
+    {reason, end_time} =
+      if bridge_movement_values.status == :raised do
+        {:chelsea_st_bridge_raised, bridge_movement_values.lowering_time}
+      else
+        {:chelsea_st_bridge_lowered, db_record.created_at + @notification_expiration_threshold}
```

Comment:
This is probably the best approach is `end_time` is a required field, though it does make me think that perhaps it ought to be optional, since this is a clear example where it doesn't make a ton of sense and we're basically just replicating the expiration logic.

Reply:
Good point, and I notice that we only actually do use it for bridge raises...will make it optional.